File: app/src/main/java/com/xenione/libs/swipemaker/Anchors.java
package com.xenione.libs.swipemaker;

import java.util.Arrays;

/**
 * Created by Eugeni on 16/04/2016.
 */
public class Anchors {

    private static class AnchorHelper {

        private final Integer[] mAnchors;

        private AnchorHelper(final Integer[] anchor) {
            super();
            mAnchors = anchor;
        }

        private int next(int index) {
            return mAnchors[index + 1];
        }

        private int prev(int index) {
            return mAnchors[index - 1];
        }

        private int pos(int index) {
            return mAnchors[index];
        }

        private int sectionFromInf(int position) {
            int section = 0;
            for (int i = 0; i < mAnchors.length - 1; i++) {
                if (mAnchors[i] > position) {
                    section = i;
                    break;
                }
            }

            return section;
        }

        private int sectionFromSup(int position) {
            int section = 0;
            for (int i = mAnchors.length - 1; i >= 0; i--) {
                if (mAnchors[i] < position) {
                    section = i;
                    break;
                }
            }

            return section;
        }

        private int size() {
            return mAnchors.length;
        }

        private int getSupLimit() {
            return mAnchors[mAnchors.length - 1];
        }

        private int getInfLimit() {
            return mAnchors[0];
        }

        private static int distance(int init, int end) {
            return Math.abs(init - end);
        }
    }

    private final AnchorHelper mAnchorHelper;

    private Anchors(final Integer[] anchor) {
        mAnchorHelper = new AnchorHelper(anchor);
    }

    public static Anchors make(Integer[] anchors) {
        if (anchors.length < 2) {
            throw new IllegalArgumentException(
                    "Amount of anchor points provided to SwipeLayout have to be bigger than 2"
            );
        }
        Arrays.sort(anchors);
        return new Anchors(anchors);
    }

    /**
     * Gives the amount of anchors that are set.
     *
     * @return amount of anchors.
     */
    public int size() {
        return mAnchorHelper.size();
    }

    /**
     * Gives distance relative (1%) within the superior limit and the
     * Inferior limit from the inferior limit.
     *
     * @param x point from where to get relative distance
     * @return 1% from inf limit
     */
    public float distance(int x) {
        return distance(x, mAnchorHelper.getSupLimit(), mAnchorHelper.getInfLimit());
    }

    /**
     * Gives distance relative (1%) within the superior Section limit and the
     * Inferior Section limit from the inferior limit.
     *
     * @param x point from where to get relative distance
     * @return 1% from inf limit
     */

    public float distance(int section, int x) {
        return distance(x, mAnchorHelper.next(section), mAnchorHelper.pos(section));
    }

    private float distance(int x, int limitSup, int limitInf) {
        return (float) (x - limitInf) / (limitSup - limitInf);
    }

    public int anchorFor(int section) {
        return mAnchorHelper.pos(section);
    }

    public int sectionFor(int position) {
        if ((position > mAnchorHelper.getSupLimit()) || (position < mAnchorHelper.getInfLimit())) {
            throw new IllegalArgumentException("position exceed limits");
        }
        return mAnchorHelper.sectionFromSup(position);
    }

    public int closeTo(int section, int point) {
        int distInf = AnchorHelper.distance(point, mAnchorHelper.pos(section));
        int distSup = AnchorHelper.distance(point, mAnchorHelper.next(section));
        if (distInf < distSup) {
            return mAnchorHelper.pos(section);
        }
        return mAnchorHelper.next(section);
    }

    public int cropInLimits(int x) {
        int inBounds = x;
        if (x < mAnchorHelper.getInfLimit()) {
            inBounds = mAnchorHelper.getInfLimit();
        } else if (x > mAnchorHelper.getSupLimit()) {
            inBounds = mAnchorHelper.getSupLimit();
        }
        return inBounds;
    }
}

File: app/src/main/java/com/xenione/libs/swipemaker/Anchors.kt
package com.xenione.libs.swipemaker

import java.util.*
import kotlin.math.abs

/**
 * Created by Eugeni on 16/04/2016.
 */
class Anchors private constructor(anchor: Array<Int>) {

    private class AnchorHelper(private val mAnchors: Array<Int>) {

        fun next(index: Int): Int {
            return mAnchors[index + 1]
        }

        fun pos(index: Int): Int {
            return mAnchors[index]
        }

        fun sectionFromSup(position: Int): Int {
            var section = 0
            for (i in mAnchors.indices.reversed()) {
                if (mAnchors[i] < position) {
                    section = i
                    break
                }
            }
            return section
        }

        fun size(): Int {
            return mAnchors.size
        }

        val supLimit: Int
            get() = mAnchors[mAnchors.size - 1]
        val infLimit: Int
            get() = mAnchors[0]

        companion object {
            fun distance(init: Int, end: Int): Int {
                return abs(init - end)
            }
        }
    }

    private val mAnchorHelper: AnchorHelper

    /**
     * Gives the amount of anchors that are set.
     *
     * @return amount of anchors.
     */
    fun size(): Int {
        return mAnchorHelper.size()
    }

    /**
     * Gives distance relative (1%) within the superior limit and the
     * Inferior limit from the inferior limit.
     *
     * @param x point from where to get relative distance
     * @return 1% from inf limit
     */
    fun distance(x: Int): Float {
        return distance(x, mAnchorHelper.supLimit, mAnchorHelper.infLimit)
    }

    /**
     * Gives distance relative (1%) within the superior Section limit and the
     * Inferior Section limit from the inferior limit.
     *
     * @param x point from where to get relative distance
     * @return 1% from inf limit
     */
    fun distance(section: Int, x: Int): Float {
        return distance(x, mAnchorHelper.next(section), mAnchorHelper.pos(section))
    }

    private fun distance(x: Int, limitSup: Int, limitInf: Int): Float {
        return (x - limitInf).toFloat() / (limitSup - limitInf)
    }

    fun anchorFor(section: Int): Int {
        return mAnchorHelper.pos(section)
    }

    fun sectionFor(position: Int): Int {
        require(!(position > mAnchorHelper.supLimit || position < mAnchorHelper.infLimit)) { "position exceed limits" }
        return mAnchorHelper.sectionFromSup(position)
    }

    fun closeTo(section: Int, point: Int): Int {
        val distInf = AnchorHelper.distance(point, mAnchorHelper.pos(section))
        val distSup = AnchorHelper.distance(point, mAnchorHelper.next(section))
        return if (distInf < distSup) {
            mAnchorHelper.pos(section)
        } else mAnchorHelper.next(section)
    }

    fun cropInLimits(x: Int): Int {
        var inBounds = x
        if (x < mAnchorHelper.infLimit) {
            inBounds = mAnchorHelper.infLimit
        } else if (x > mAnchorHelper.supLimit) {
            inBounds = mAnchorHelper.supLimit
        }
        return inBounds
    }

    companion object {
        @JvmStatic
        fun make(anchors: Array<Int>): Anchors {
            require(anchors.size >= 2) { "Amount of anchor points provided to SwipeLayout have to be bigger than 2" }
            Arrays.sort(anchors)
            return Anchors(anchors)
        }
    }

    init {
        mAnchorHelper = AnchorHelper(anchor)
    }
}